## Ejemplos de visualizaciones con librería Plotly.

Cargamos unos datos de estadísticas avanzadas de esta temporada de la NBA, y vamos a realizar ejemplos de muchas de las visualizaciones que tiene la librería Plotly.

In [1]:
# imports necesarios
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

import plotly.express as px
import plotly.figure_factory as ff
import plotly.graph_objects as go
import plotly.offline as po
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot
from plotly.subplots import make_subplots

import random

In [2]:
# carga de datos
df_nba_advanced = pd.read_csv("NBA_2020_advanced.csv")
df_nba_advanced.head(10)

Player Pos  Age   Tm   G    MP   PER    TS%   3PAr  \
0              Steven Adams   C   26  OKC  38  1052  22.4  0.622  0.003   
1               Bam Adebayo   C   22  MIA  41  1407  20.7  0.624  0.021   
2         LaMarcus Aldridge   C   34  SAS  38  1250  20.7  0.585  0.183   
3  Nickeil Alexander-Walker  SG   21  NOP  33   453   8.8  0.451  0.524   
4             Grayson Allen  SG   24  MEM  27   461  11.8  0.589  0.528   
5             Jarrett Allen   C   21  BRK  40  1064  21.1  0.670  0.011   
6              Kadeem Allen  SG   27  NYK   9    96  13.6  0.483  0.381   
7           Al-Farouq Aminu  PF   29  ORL  18   380   7.8  0.395  0.419   
8           Justin Anderson  SF   26  BRK   3    17  -3.7  0.218  0.500   
9             Kyle Anderson  PF   26  MEM  36   603  12.6  0.503  0.217   

     FTr  ...  USG%  OWS  DWS   WS  WS/48  OBPM  DBPM   BPM  VORP  Year  
0  0.422  ...  17.4  3.0  1.6  4.6  0.210   2.1   3.3   5.4   2.0  2020  
1  0.533  ...  19.9  3.2  2.3  5.5  0.189   1.1   3.8   5.0   2.5  2020  
2  0.230  ...  23.0  2.8  1.0  3.8  0.145   1.6   0.5   2.1   1.3  2020  
3  0.115  ...  22.9 -0.5  0.2 -0.3 -0.027  -3.2  -2.5  -5.7  -0.4  2020  
4  0.202  ...  18.1  0.5  0.1  0.6  0.064  -0.4  -3.0  -3.4  -0.2  2020  
5  0.588  ...  15.3  3.0  1.8  4.8  0.219   0.4   2.8   3.1   1.4  2020  
6  0.190  ...  23.5  0.0  0.0  0.0  0.012  -1.4  -3.1  -4.5  -0.1  2020  
7  0.337  ...  13.2 -0.4  0.6  0.2  0.025  -4.8   2.5  -2.3   0.0  2020  
8  0.333  ...  16.9 -0.1  0.0 -0.1 -0.168 -12.7  -2.6 -15.3  -0.1  2020  
9  0.270  ...  13.7  0.4  0.6  1.0  0.077  -1.9   1.4  -0.5   0.2  2020  

[10 rows x 27 columns]

In [3]:
df_nba_advanced.columns

Index(['Player', 'Pos', 'Age', 'Tm', 'G', 'MP', 'PER', 'TS%', '3PAr', 'FTr',
       'ORB%', 'DRB%', 'TRB%', 'AST%', 'STL%', 'BLK%', 'TOV%', 'USG%', 'OWS',
       'DWS', 'WS', 'WS/48', 'OBPM', 'DBPM', 'BPM', 'VORP', 'Year'],
      dtype='object')

###### Datasets estadísticas avanzadas
- Player -- Jugador
- Pos -- Position
- Age -- Age of Player at the start of February 1st of that season.
- Tm -- Team
- G -- Games
- MP -- Minutes Played
- PER -- Player Efficiency Rating. Una medida de producción por minuto estandarizada de modo que el promedio de la liga sea 15.
- TS% -- True Shooting Percentage. Porcentaje de tiro real. Una medida de eficiencia de tiro que tiene en cuenta los tiros de 2 puntos, los tiros de 3 puntos y los tiros libres.
- 3PAr -- 3-Point Attempt Rate. Porcentaje de intentos de FG desde el rango de 3 puntos.
- FTr -- Free Throw Attempt Rate. Número de intentos de FT por intento de FG.
- ORB% -- Offensive Rebound Percentage. Porcentaje de rebote ofensivo. Una estimación del porcentaje de rebotes ofensivos disponibles que un jugador agarró mientras estaba jugando.
- DRB% -- Defensive Rebound Percentage. Porcentaje de rebote defensivo. Una estimación del porcentaje de rebotes defensivos disponibles que un jugador agarró mientras estaba jugando.
- TRB% -- Total Rebound Percentage. Porcentaje de rebote total. Una estimación del porcentaje de rebotes disponibles que un jugador agarró mientras estaba jugando.
- AST% -- Assist Percentage. Porcentaje de asistencia. Una estimación del porcentaje de canastas de compañeros de equipo a los que ayudó un jugador mientras estaba jugando.
- STL% -- Steal Percentage. Porcentaje de robo. Una estimación del porcentaje de posesiones del oponente que termina con un robo por parte del jugador mientras estaba jugando.
- BLK% -- Block Percentage. Porcentaje de bloqueo. Una estimación del porcentaje de intentos de tiros de campo de dos puntos del oponente bloqueados por el jugador mientras estaba jugando.
- TOV% -- Turnover Percentage. Porcentaje de pérdidas. Una estimación de pérdidas de balón cometidas por cada 100 jugadas.
- USG% -- Usage Percentage. Porcentaje de uso. Una estimación del porcentaje de jugadas de equipo utilizadas por un jugador mientras estaba jugando.
- OWS -- Offensive Win Shares. Acciones ofensivas de victoria. Una estimación del número de victorias aportadas por un jugador debido a su ataque.
- DWS -- Defensive Win Shares. Acciones de victoria defensivas. Una estimación del número de victorias aportadas por un jugador debido a su defensa.
- WS -- Win Shares. Una estimación del número de victorias aportadas por un jugador.
- WS/48 -- Win Shares Per 48 Minutes. Una estimación del número de victorias aportadas por un jugador por 48 minutos (el promedio de la liga es aproximadamente .100)
- OBPM -- Offensive Box Plus/Minus. Una estimación de puntuación mas/menos de los puntos ofensivos por 100 posesiones que un jugador contribuyó por encima de un jugador promedio de la liga, traducido a un equipo promedio.
- DBPM -- Defensive Box Plus/Minus. Una estimación de puntuación mas/menos de los puntos defensivos por 100 posesiones que un jugador contribuyó por encima de un jugador promedio de la liga, traducido a un equipo promedio.
- BPM -- Box Plus/Minus. Una estimación de puntuación mas/menos de los puntos por 100 posesiones que un jugador contribuyó por encima de un jugador promedio de la liga, traducido a un equipo promedio.
- VORP -- Value over Replacement Player. Valor sobre el jugador de reemplazo. Una estimación de puntuación mas/menos de los puntos por 100 posesiones de EQUIPO que un jugador contribuyó por encima de un jugador de nivel de reemplazo (-2.0), traducido a un equipo promedio y prorrateado a una temporada de 82 juegos. Multiplique por 2.70 para convertir a victorias sobre reemplazo.
- Year -- Año/temporada al que hacen referencia las estadísticas.

In [5]:
# scatter plot entre variables PER y OBPM
data = [go.Scatter(x = df_nba_advanced.PER, y = df_nba_advanced.OBPM, mode = 'markers',
                     marker = dict(size = 7, color = '#4ED700', symbol = 'circle',
                              line = dict(width = 1,color = '#0E8700')))]

layout = go.Layout(title = '$NBA Stats$', xaxis = dict(title = '$PER$'), yaxis = dict(title = '$OBPM$'))

fig = go.Figure(data=data, layout=layout)

iplot(fig)

In [7]:
# scatter plot añadiendo al eje y dos variables

OBPM = go.Scatter( x = df_nba_advanced.PER, y = df_nba_advanced.OBPM, mode = 'markers',
                    marker = dict(size = 7, color = '#F4D03F', symbol = 'circle',
                                  line = dict(width = 1,color = '#0E8700')))

DBPM = go.Scatter( x = df_nba_advanced.PER, y = df_nba_advanced.DBPM, mode = 'markers',
                    marker = dict(size = 7, color = '#A9DFBF', symbol = 'circle',
                                  line = dict(width = 1,color = '#0E8700')))

layout = go.Layout( title = '$NBA Stats$', xaxis = dict(title = '$PER$'), 
                    yaxis = dict(title = '$OBPM/DBPM$'))

data = [OBPM, DBPM]  
fig = go.Figure(data=data, layout=layout)
iplot(fig)

In [6]:
# trace0 con los datos de los jugadores de Boston, y trace1 con los datos de los jugadores de Lakers
trace0 = go.Scatter( x = df_nba_advanced[df_nba_advanced.Tm == 'BOS'].PER, 
                     y = df_nba_advanced[df_nba_advanced.Tm == 'BOS'].OBPM,
                     mode = 'markers', name = 'Celtics',
                     marker = dict(size = 7, color = '#F39C12',symbol = 'circle',line = dict(width = 1)))

trace1 = go.Scatter( x = df_nba_advanced[df_nba_advanced.Tm == 'LAL'].PER,
                     y = df_nba_advanced[df_nba_advanced.Tm == 'LAL'].OBPM,
                     mode = 'markers', name = 'Lakers',
                     marker = dict(size = 7, color = '#8BC34A',symbol = 'circle',line = dict(width = 1)))

layout = go.Layout( title = '$Scatter Plot$', xaxis = dict(title = '$PER$'), yaxis = dict(title = '$OBPM$'))

data = [trace0, trace1]  
fig = go.Figure(data=data,layout=layout)
iplot(fig)

In [9]:
# trace0 datos de Boston, trace1 datos de Lakers, y cambiamos el estilo para ver otras posibilidades.
trace0 = go.Scatter( x = df_nba_advanced[df_nba_advanced.Tm == 'BOS'].PER,
                     y = df_nba_advanced[df_nba_advanced.Tm == 'BOS'].OBPM,
                     mode = 'markers', name = 'Celtics',
                     marker = dict(size = 14, color = '#9ca4c4',symbol = 'circle',
                     line = dict(width = 1,color = '#5D6D7E')))

trace1 = go.Scatter( x = df_nba_advanced[df_nba_advanced.Tm == 'LAL'].PER,
                     y = df_nba_advanced[df_nba_advanced.Tm == 'LAL'].OBPM,
                     mode = 'markers', name = 'Lakers',
                     marker = dict(size = 14, color = '#cbcbcb',symbol = 'circle',
                     line = dict(width = 1,color = '#626567')))

layout = go.Layout( title = dict(text = "NBA Data", x=0.5, y=0.95),
                    title_font_size = 30, title_font_color = '#F1C40F',
                    xaxis = dict(showgrid=False, showline=True, linecolor='black', tickfont_color='black', 
                                showticklabels=True, dtick=5, ticks='outside', tickcolor='black'),
                    yaxis = dict(showgrid=False, showline=True, linecolor='black', tickfont_color='black',
                                showticklabels=True, ticks='outside', tickcolor='black'),
                    legend = dict(font_size=15, yanchor='bottom', xanchor='right'),
                    paper_bgcolor='white', plot_bgcolor='white', hovermode='closest', width=900, height=700)

data = [trace0, trace1]  
fig = go.Figure(data=data,layout=layout)
iplot(fig)

In [13]:
# mostramos los datos de cada equipo en un subplot distinto
fig = make_subplots(rows=1, cols=2, subplot_titles=("Boston", "Lakers"))

fig.add_trace(go.Scatter
                 (x = df_nba_advanced[df_nba_advanced.Tm == 'BOS'].PER,
                  y = df_nba_advanced[df_nba_advanced.Tm == 'BOS'].OBPM, 
                  mode = 'markers',
                  marker = dict(size = 10, color = 'crimson',symbol = 'circle',
                        line = dict(width = 1,color = '#0E8700'))),
                  row=1, col=1)

fig.add_trace(go.Scatter
                 (x = df_nba_advanced[df_nba_advanced.Tm == 'LAL'].PER,
                  y = df_nba_advanced[df_nba_advanced.Tm == 'LAL'].OBPM, 
                  mode = 'markers',
                  marker = dict(size = 10, color = 'gold',symbol = 'circle',line = dict(width = 1))),
                  row=1, col=2)

fig.show()

In [15]:
# uso de plotly express
fig = px.scatter( df_nba_advanced, 
                  x =df_nba_advanced.PER,
                  y = df_nba_advanced['OBPM'],
                  color="OBPM", 
                  size=df_nba_advanced['USG%'], # la variable no puede tomar valores negativos
                  color_discrete_map={"yes": "#FF5722","no": "#7CB342"})

fig.show()

In [20]:
# scatter plot entre PER y TS%, separando por posición
fig = px.scatter(df_nba_advanced, 
                 x=df_nba_advanced.PER,
                 y= df_nba_advanced['TS%'],
                 color="TS%", facet_col=df_nba_advanced['Pos'], 
                 color_discrete_map={"yes": "#FF5722","no": "#7CB342"}, 
                 width=980, height=500, title="PER-TS% por posición")

fig.show()

In [25]:
# además de usar ver el PER-TS% por posición, también usamos la variable Age como otra facet
fig = px.scatter(df_nba_advanced, 
                 x=df_nba_advanced.PER,
                 y= df_nba_advanced['TS%'],
                 color="TS%",   
                 facet_col=df_nba_advanced['Pos'], facet_row=df_nba_advanced['Age'], 
                 color_discrete_map={"yes": "#FF5722","no": "#7CB342"},
                 width=950, height=1800, title="PER-TS% por posición y edad")

fig.show()

In [29]:
# Gráficos Plot.
OBPM = go.Scatter( x = df_nba_advanced.index.values,
                   y = df_nba_advanced['OBPM'],
                   mode = 'lines', name = 'OBPM')

DBPM = go.Scatter( x = df_nba_advanced.index.values,
                   y = df_nba_advanced['DBPM'],
                   mode = 'lines', name = 'DBPM')

PER = go.Scatter( x = df_nba_advanced.index.values,
                  y = df_nba_advanced['PER'],
                  mode = 'lines', name = 'PER')

layout = go.Layout( title = '$Plot-OBPM-DBPM-PER$',
                    xaxis = dict(title = 'Jugadores'), 
                    yaxis = dict(title = 'Valores atributos'))

data = [OBPM, DBPM, PER]  
fig = go.Figure(data=data,layout=layout)
iplot(fig)

In [30]:
# formamos un dataframe más pequeño, solo con los datos de los jugadores de los Boston Celtics
df_boston = df_nba_advanced[df_nba_advanced['Tm'] == 'BOS']
df_boston.head()

Player Pos  Age   Tm   G    MP   PER    TS%   3PAr    FTr  ...  \
67     Jaylen Brown  SF   23  BOS  35  1169  18.0  0.598  0.359  0.302  ...   
139  Carsen Edwards  SG   21  BOS  27   267   6.5  0.432  0.621  0.137  ...   
149      Tacko Fall   C   24  BOS   4    21  24.1  0.690  0.000  0.273  ...   
185   Javonte Green  SF   26  BOS  32   248  14.0  0.623  0.235  0.221  ...   
205  Gordon Hayward  SF   29  BOS  24   759  18.4  0.596  0.319  0.141  ...   

     USG%  OWS  DWS   WS  WS/48  OBPM  DBPM   BPM  VORP  Year  
67   24.2  1.9  1.7  3.5  0.145   1.5  -0.1   1.3   1.0  2020  
139  18.2 -0.3  0.3  0.1  0.009  -3.8  -1.0  -4.8  -0.2  2020  
149  31.1  0.0  0.1  0.0  0.078 -10.7  -4.1 -14.8  -0.1  2020  
185  14.7  0.3  0.4  0.7  0.136  -1.7   0.5  -1.2   0.0  2020  
205  20.2  1.7  1.0  2.6  0.166   2.1   0.4   2.6   0.9  2020  

[5 rows x 27 columns]

In [34]:
# Plot como el anterior pero solo de los jugadores de Boston, y poniendo en el eje x los nombres de jugadores
OBPM = go.Scatter(x = df_boston.Player.values,
                  y = df_boston['OBPM'],
                  mode = 'lines', name = 'OBPM')

DBPM = go.Scatter(x = df_boston.Player.values,
                  y = df_boston['DBPM'],
                  mode = 'lines', name = 'DBPM')

PER = go.Scatter(x = df_boston.Player.values,
                 y = df_boston['PER'],
                 mode = 'lines', name = 'PER')

layout = go.Layout(title = '$PLOT BOSTON - OBPM - DBPM - PER$',
                    xaxis = dict(title = 'Jugadores'), 
                    yaxis = dict(title = 'Valores atributos'))

data = [OBPM, DBPM, PER]  
fig = go.Figure(data=data,layout=layout)
iplot(fig)

In [37]:
# usamos bucle for para hacer el plot con varios atributos ahorrando código
traces = []

# nos quedamos con las columnas que nos interesen, por no tener tantos atributos que impiden visualizar bien
for i in df_boston.columns:
    if i in('OWS','DWS','WS','OBPM','DBPM','BPM','VORP'):
        traces.append(
            go.Scatter( x=df_boston.Player.values,
                        y=df_boston[i],
                        mode='lines', name = i, connectgaps=True)) 
    
layout = go.Layout( title = 'PLOT-BOSTON-FOR', title_font=dict(size=20),
                    xaxis = dict(title = 'Jugadores'), yaxis = dict(title = 'Valores atributos'), 
                    font=dict(size=10), width=1000, height=650)

fig = go.Figure(data=traces, layout=layout)
fig.show()

In [39]:
# dos plots juntos, representando distintos datos.

fig = make_subplots(rows=1, cols=2, subplot_titles=("OBPM-DBPM-PER", "PER-OBPM-BOSTON-LAKERS"))

fig.add_trace(go.Scatter(x = df_nba_advanced.index.values, y = df_nba_advanced['OBPM'],
                    mode = 'lines', name = 'OBPM'),
                    row=1, col=1)

fig.add_trace(go.Scatter(x = df_nba_advanced.index.values, y = df_nba_advanced['DBPM'],
                    mode = 'lines', name = 'DBPM'),
                    row=1, col=1)

fig.add_trace(go.Scatter(x = df_nba_advanced.index.values, y = df_nba_advanced['PER'],
                    mode = 'lines', name = 'PER'),
                    row=1, col=1)

fig.add_trace(go.Scatter(x = df_nba_advanced[df_nba_advanced.Tm == 'BOS'].PER,
                         y = df_nba_advanced[df_nba_advanced.Tm == 'BOS'].OBPM,
                         mode = 'markers', name = 'Boston',
                         marker = dict(size = 7, color = '#F39C12',symbol = 'circle',line = dict(width = 1))), 
                         row=1, col=2)

fig.add_trace(go.Scatter(x = df_nba_advanced[df_nba_advanced.Tm == 'LAL'].PER,
                         y = df_nba_advanced[df_nba_advanced.Tm == 'LAL'].OBPM,
                         mode = 'markers',
                         name = 'Lakers',
                         marker = dict(size = 7, color = '#8BC34A',symbol = 'circle',line = dict(width = 1))), 
                         row=1, col=2)

fig.update_xaxes(title_text="Jugadores", row=1, col=1)
fig.update_yaxes(title_text="Valores atributos", row=1, col=1)
fig.update_xaxes(title_text="PER" ,row=1, col=2)
fig.update_yaxes(title_text="OBPM", row=1, col=2)

fig.update_layout( paper_bgcolor= '#FFFDE7', plot_bgcolor= '#FFFDE7',
                   title=dict(text = "Sub Plots",x=0.5,y=0.95), title_font_size=30)

fig.show()

In [42]:
# Uso de plotly express.
fig = px.line(x=df_nba_advanced.index.values, y=df_nba_advanced['PER'])

fig.add_scatter(x=df_nba_advanced.index.values, y=df_nba_advanced['OBPM'], 
                 name = 'OBPM', line={'color': 'green'})

fig.add_scatter(x=df_nba_advanced.index.values, y= df_nba_advanced['DBPM'] , 
                 name = 'DBPM', line={'color': 'red'})

fig.update_layout(title_text='PER-OBPM-DBPM')
fig.show()

In [43]:
# plotline con la variable PER en el eje y, y las lineas de colores representando la posición del jugador
fig = px.line(df_nba_advanced, x = df_nba_advanced.index.values, y = "PER", color = 'Pos')

fig.update_xaxes(showgrid=False)
fig.update_yaxes(showgrid=False)

fig.show()

In [44]:
# gráfica de barras horizontales, con los jugadores de Boston y sus valores de PER
x = df_boston.PER.values
y = df_boston.Player.values

data = go.Bar( x = x, y = y, marker={'color' : '#FFA726'}, orientation='h')

layout = go.Layout( title = 'Simple Bar Chart', width=970, height=650)

fig = go.Figure(data=data,layout=layout)
fig.show()

In [49]:
# Gráfica de barras, con otro estilo, mostrando los jugadores y su PER
# datos ordenados por PER
df_sort = df_boston.sort_values(by = 'PER')

fig = go.Figure()

fig.add_trace(go.Bar(x = df_sort.PER, y = df_sort.Player,
                     orientation = 'h', marker = dict(color='rgba(70, 171, 100, 0.6)',
                     line = dict(color='rgba(70, 171, 100, 1.0)',width=1))))

fig.update_layout( title = dict(text = "Boston Celctis PER by Player",x=0.46,y=0.95,font_size=20),
                   yaxis = dict(showticklabels=True),
                   xaxis = dict(showticklabels=False, domain=[0, 0.6]),
                   margin = dict(l=300, r=20, t=70, b=70),
                   paper_bgcolor = 'rgb(248, 248, 255)',
                   plot_bgcolor = 'rgb(248, 248, 255)')

annotations = [] 

for perc, asg in zip(df_sort.PER, df_sort.Player):
    annotations.append(dict(xref='x', yref='y', y=asg, x=perc + 2, text=str(perc),
                            font=dict(family='Arial', size=12,color='#004D40'), showarrow=False))
   
    annotations.append(dict(xref='paper', yref='paper', x=1, y=-0.17,
                            text='Boston Celtics ' +'Per by player, ' +'Year 2020',
                            font=dict(family='Arial', size=11, color='#9E9E9E'), showarrow=False))

fig.update_layout(annotations=annotations)
fig.show()

In [50]:
# gráfica de barras verticales mostrando dos variables para cada valor del eje x
x = df_boston.Player.values
y1 = df_boston.DBPM.values
y2 = df_boston.OBPM.values

trace0 = go.Bar( x = x, y = y1, marker = dict(color = '#FFA726'))

trace1 = go.Bar( x = x, y = y2, marker={'color' : '#94E413'})

data = [trace0,trace1]
fig = go.Figure(data=data)
fig.show()

In [52]:
# barras verticales, agrupando en una misma barra los distintos atributos para cada valor del eje x
x = df_boston.Player.values
y1 = df_boston.OWS.values
y2 = df_boston.DWS.values
y3 = df_boston.WS.values

trace0 = go.Bar(x = x, y = y1, marker = dict (color = '#FF6F00' ), name = 'OWS')

trace1 = go.Bar(x = x, y = y2, marker = {'color' : '#FFB300'}, name = 'DWS')    

trace2 = go.Bar(x = x, y = y3, marker = {'color' : '#F7DC6F'}, name = 'WS')  
    
layout = go.Layout(title = 'OWS-DWS-WS', barmode = 'stack', width=900, height=800)

data = [trace0,trace1,trace2]
fig = go.Figure(data = data, layout = layout)
fig.show()

In [54]:
# barras verticales, agrupando en una misma barra los distintos atributos para cada valor del eje x, y 
# poniendo en cada sitio el valor numérico
x = df_boston.Player.values
y1 = df_boston.OWS.values
y2 = df_boston.DWS.values
y3 = df_boston.WS.values

trace0 = go.Bar(x = x, y = y1, marker = dict (color = '#FF6F00'),
             name = 'OWS', text = y1, textposition = 'auto')

trace1 = go.Bar(x = x, y = y2, marker = {'color' : '#FFB300'},
             name = 'DWS', text = y2, textposition = 'auto')    

trace2 = go.Bar(x = x, y = y3, marker = {'color' : '#F7DC6F'},
             name = 'WS', text = y3, textposition = 'auto')  
    
layout = go.Layout(title = 'OWS-DWS-WS', barmode = 'stack',
             width=900, height=800)

data = [trace0,trace1,trace2]
fig = go.Figure(data = data, layout = layout)
fig.show()

In [58]:
# como el anterior pero en barras horizontales
x = df_boston.Player.values
y1 = df_boston.OWS.values
y2 = df_boston.DWS.values
y3 = df_boston.WS.values

trace0 = go.Bar(x = y1, y = x, marker = dict (color = '#FF6F00'),
             name = 'OWS', orientation='h', text = y1, textposition = 'auto')

trace1 = go.Bar(x = y2, y = x, marker = {'color' : '#FFB300'},
             name = 'DWS', orientation='h', text = y2, textposition = 'auto')    

trace2 = go.Bar(x = y3, y = x, marker = {'color' : '#F7DC6F'},
             name = 'WS', orientation='h', text = y3, textposition = 'auto')  
    
layout = go.Layout(title = 'OWS-DWS-WS', barmode = 'stack',
             width=990, height=550)

data = [trace0,trace1,trace2]
fig = go.Figure(data = data, layout = layout)
fig.show()

In [59]:
# como el anterior pero cambiando un poco el aspecto
x = df_boston.Player.values
y1 = df_boston.OWS.values
y2 = df_boston.DWS.values
y3 = df_boston.WS.values

trace0 = go.Bar(x = y1, y = x, marker = dict(color ='#FF9800', line=dict(color='#F4511E',width=2)),
             name = 'OWS', orientation='h', text = y1, textposition = 'auto', opacity=0.8)

trace1 = go.Bar(x = y2, y = x, marker = dict(color ='#7CB342',line=dict(color='#2E7D32', width=2)),
             name = 'DWS', orientation='h', text = y2, textposition = 'auto', opacity=0.8)    

trace2 = go.Bar(x = y3, y = x, marker = dict(color ='#1E88E5',line=dict(color='#3F51B5', width=2)),
             name = 'WS', orientation='h', text = y3, textposition = 'auto', opacity=0.7)  
    
layout = go.Layout(title = 'OWS-DWS-WS', barmode = 'stack',
             width=990, height=600)

data = [trace0,trace1,trace2]
fig = go.Figure(data = data, layout = layout)
fig.show()

In [61]:
# barras verticales, sin agruparlas en 1 sola por cada valor del eje x
x = df_boston.Player.values
y1 = df_boston.OWS.values
y2 = df_boston.DWS.values
y3 = df_boston.WS.values

trace0 = go.Bar(x = x, y = y1, marker = dict (color ='#FF6F00'), name = 'OWS')

trace1 = go.Bar(x = x, y = y2, marker = {'color' : '#FFB300'}, name = 'DWS')    

trace2 = go.Bar(x = x, y = y3, marker = {'color' : '#F7DC6F'}, name = 'WS')  
    
layout = go.Layout(title = 'OWS-DWS-WS', width = 980, height = 800)

data = [trace0,trace1,trace2]
fig = go.Figure(data = data, layout = layout)
fig.show()

In [63]:
# similar al anterior pero añadiendo el valor numérico a cada barra
x = df_boston.Player.values
y1 = df_boston.OWS.values
y2 = df_boston.DWS.values
y3 = df_boston.WS.values

trace0 = go.Bar(x = x, y = y1, marker = dict (color ='#FF6F00'), name = 'OWS', text = y1, textposition='outside')

trace1 = go.Bar(x = x, y = y2, marker = {'color' : '#FFB300'}, name = 'DWS', text = y2, textposition='outside')    

trace2 = go.Bar(x = x, y = y3, marker = {'color' : '#F7DC6F'}, name = 'WS', text = y3, textposition='outside')  
    
layout = go.Layout(title = 'OWS-DWS-WS', barmode = 'group', width = 980, height = 800, xaxis_tickangle=-45)

data = [trace0,trace1,trace2]
fig = go.Figure(data = data, layout = layout)
fig.show()

In [65]:
# similar al anterior pero cambiando la ubicación de los valores dentro de las barras
x = df_boston.Player.values
y1 = df_boston.OWS.values
y2 = df_boston.DWS.values
y3 = df_boston.WS.values

trace0 = go.Bar(x = x, y = y1, marker = dict (color ='#FF6F00'), name = 'OWS', text = y1, textposition='auto')

trace1 = go.Bar(x = x, y = y2, marker = {'color' : '#FFB300'}, name = 'DWS', text = y2, textposition='auto')    

trace2 = go.Bar(x = x, y = y3, marker = {'color' : '#F7DC6F'}, name = 'WS', text = y3, textposition='auto')  

layout = go.Layout(title = dict(text = "OWS-DWS-WS'",x=0.5,y=0.95,font_size=25), 
                   barmode = 'group',
                   width = 980, height = 800,
                   xaxis_tickangle = -45, xaxis_tickfont_size = 14,
                   yaxis = dict(title='Valores atributos', titlefont_size=16, tickfont_size=14),
                   xaxis = dict(title='Jugadores', titlefont_size=16, tickfont_size=14), 
                   legend = dict(x=0, y=1, bgcolor='rgba(255, 255, 255, 0)', bordercolor='rgba(255, 255, 255, 0)'),
                   bargap = 0.15, 
                   bargroupgap=0.08)

layout = go.Layout(title = 'OWS-DWS-WS', barmode = 'group', width = 980, height = 800, xaxis_tickangle=-45)

data = [trace0,trace1,trace2]
fig = go.Figure(data = data, layout = layout)
fig.show()

In [68]:
# uso de plotly express para gráfica de barras
x = df_boston.Player.values
y = df_boston.PER.values

fig = px.bar(x = x, y = y, color = y, text = y)
fig.layout.title.text = "Bar Plot - Plotly Express"
fig.update_traces(textposition='outside')
fig.show()

In [72]:
# barras agrupadas con varias variables de forma horizontal
fig = go.Figure()

fig.add_trace(go.Bar(x = df_boston['TS%'], y = df_boston.Player.values,
                       marker = dict(color ='#2E7D32',line = dict(color='#2E7D32',width=2)),
                       name = 'TS%', orientation = 'h', textposition = 'auto'))

fig.add_trace(go.Bar(x = df_boston['TRB%'], y = df_boston.Player.values,
                       marker = dict(color ='#8BC34A',line=dict(color='#8BC34A',width=2)),
                       name = 'TRB%', orientation = 'h', textposition = 'auto'))

fig.add_trace(go.Bar(x = df_boston['AST%'], y = df_boston.Player.values,
                       marker = dict (color ='#D4E157',line=dict(color='#D4E157',width=2)),
                       name = 'AST%', orientation = 'h', textposition = 'auto'))

fig.add_trace(go.Bar(x = df_boston['BLK%'], y = df_boston.Player.values,
                       marker= dict (color ='#FFB300',line=dict(color='#FFB300',width=2)),
                       name = 'BLK%', orientation='h', textposition='auto'))

fig.add_trace(go.Bar(x = df_boston['STL%'], y = df_boston.Player.values,
                       marker = dict (color ='#FF7043',line=dict(color='#FF7043',width=2)),
                       name = 'STL%', orientation ='h', textposition ='auto'))

fig.update_layout(title = dict(text = "Boston Celtics",x=0.44,y=0.95,font_size=20), 
                   barmode = 'stack', width = 950, height = 500,
                   margin = dict(l=70, r=0, t=70, b=70),
                   paper_bgcolor = 'rgb(248, 248, 255)', plot_bgcolor = 'rgb(248, 248, 255)')


fig.update_layout()
fig.show()

In [79]:
# Pie chart minutos por posición
minutos = df_boston.MP.values
posicion = df_boston.Pos.values
colors = ['#8BC34A','#D4E157','#FFB300','#FF7043']

data = go.Pie(values = minutos, labels = posicion,
              marker = dict(colors=colors), textinfo = 'label+value+percent')

layout = go.Layout(title = dict(text = "Minutos por posición",x=0.46,y=0.95,font_size=20),
                   width=800, height=650)

fig = go.Figure(data=data,layout=layout)
fig.show()

In [80]:
# Similar al anterior, pero mostrando información al pasar por encima y cambiando la orientación del texto
minutos = df_boston.MP.values
posicion = df_boston.Pos.values
colors = ['#8BC34A','#D4E157','#FFB300','#FF7043']

data = go.Pie(values = minutos, labels = posicion,
              marker = dict(colors = colors,line = dict(color='#797D7F', width=1.5)),
              textinfo = 'percent', hoverinfo = 'label+value',
              textfont_size = 15, insidetextorientation = 'radial')

layout = go.Layout(title = dict(text = "Minutos por posición", x=0.46, y=0.95, font_size=20),
                   width = 800, height = 650)

fig = go.Figure(data=data,layout=layout)
fig.show()

In [85]:
# Mostramos dos gráficas pie en una imagen
minutos = df_boston.MP.values
posicion = df_boston.Pos.values
colors = ['#8BC34A','#D4E157','#FFB300','#FF7043']
player = df_boston.Player.values
minutos = df_boston.MP.values

#Subplot initialization
fig = make_subplots(rows=1, cols=2, subplot_titles=("Minutos por posición", "Minutos por jugador"),
                    specs=[[{'type':'domain'}, {'type':'domain'}]])

fig.add_trace(go.Pie(values = minutos, labels = posicion,
                     marker = dict(colors=colors), textinfo = 'label+value+percent'),
                     row = 1, col = 1)

fig.add_trace(go.Pie(values= minutos, labels= player,
                     marker=dict(colors=colors), textinfo='label+value+percent'),
                     row=1, col=2)

fig.update_layout(paper_bgcolor= '#FFFDE7', plot_bgcolor= '#FFFDE7', 
                  title=dict(text = "Pie charts", x=0.5, y=0.95), title_font_size=30,
                   width = 1200, height = 750)

fig.show()

In [86]:
# Mostramos dos gráficas donut en una imagen
minutos = df_boston.MP.values
posicion = df_boston.Pos.values
colors = ['#8BC34A','#D4E157','#FFB300','#FF7043']
player = df_boston.Player.values
minutos = df_boston.MP.values

#Subplot initialization
fig = make_subplots(rows=1, cols=2, subplot_titles=("Minutos por posición", "Minutos por jugador"),
                    specs=[[{'type':'domain'}, {'type':'domain'}]])

fig.add_trace(go.Pie(values = minutos, labels = posicion, hole = .4,
                     marker = dict(colors=colors), textinfo = 'label+value+percent', hoverinfo='label',), 
                     row = 1, col = 1)

fig.add_trace(go.Pie(values= minutos, labels= player, hole = .4,
                     marker=dict(colors=colors), textinfo='label+value+percent', hoverinfo='label',), 
                     row=1, col=2)

fig.update_layout(paper_bgcolor= '#FFFDE7', plot_bgcolor= '#FFFDE7',
                  title=dict(text = "Pie charts", x=0.5, y=0.95), title_font_size=30,
                   width = 1200, height = 750)

fig.show()

In [89]:
# ejemplo Pie con plotly express
fig = px.pie(df_boston, values = 'MP', names = 'Pos', title='Pie chart')

fig.update_layout(title=dict(text = "Minutos por posición",x=0.44,y=0.95,font_size=20),
                   width=800, height=650)

fig.update_traces(textposition = 'outside', textinfo = 'percent+label')
fig.show()

In [90]:
# Area plot per by player
x = df_boston.Player.values
y = df_boston.PER.values

fig = go.Figure()
fig.add_trace(go.Scatter(x = x, y = y, fill='tozeroy',marker = dict(color = 'gold')))
fig.show()

In [91]:
# gráfico de áreas con varios atributos
x = df_boston.Player.values
y1 = df_boston['TRB%'].values
y2 = df_boston['AST%'].values
y3 = df_boston['BLK%'].values
y4 = df_boston['STL%'].values

fig = go.Figure()
fig.add_trace(go.Scatter(x=x, y=y1, fill='tozeroy',marker = dict(color = '#00b159'))) 
fig.add_trace(go.Scatter(x=x, y=y2, fill='tonexty',marker = dict(color = '#ffc425')))
fig.add_trace(go.Scatter(x=x, y=y3, fill='tonexty',marker = dict(color = '#3498DB'))) 
fig.add_trace(go.Scatter(x=x, y=y4, fill='tonexty',marker = dict(color = '#ff3b30'))) 
fig.update_layout(width = 980 , height = 600)

fig.show()

In [92]:
# Stacked Area Chart. Cada area se va poniendo seguida a la siguiente, en lugar de sobreponerse unas a otras.
x = df_boston.Player.values
y1 = df_boston['TRB%'].values
y2 = df_boston['AST%'].values
y3 = df_boston['BLK%'].values
y4 = df_boston['STL%'].values

fig = go.Figure()
fig.add_trace(go.Scatter(x=x, y=y1, marker = dict(color = '#00b159'), stackgroup='one'))
fig.add_trace(go.Scatter(x=x, y=y2, marker = dict(color = '#ffc425'), stackgroup='one'))
fig.add_trace(go.Scatter(x=x, y=y3, marker = dict(color = '#3498DB'), stackgroup='one'))
fig.add_trace(go.Scatter(x=x, y=y4, marker = dict(color = '#ff3b30'), stackgroup='one'))

fig.update_layout(width = 980 , height = 600)
fig.show()

In [94]:
# similar al anterior cambiando estilo
x = df_boston.Player.values
y1 = df_boston['TRB%'].values
y2 = df_boston['AST%'].values
y3 = df_boston['BLK%'].values
y4 = df_boston['STL%'].values

fig = go.Figure()
fig.add_trace(go.Scatter(x=x, y=y1, mode='none', stackgroup='one'))
fig.add_trace(go.Scatter(x=x, y=y2, mode='none', stackgroup='one'))
fig.add_trace(go.Scatter(x=x, y=y3, mode='none', stackgroup='one'))
fig.add_trace(go.Scatter(x=x, y=y4, mode='none', stackgroup='one'))

fig.update_layout(width = 980 , height = 600)
fig.show()

In [95]:
# similar a anteriores pero ocupando el 100% del area disponible
x = df_boston.Player.values
y1 = df_boston['TRB%'].values
y2 = df_boston['AST%'].values
y3 = df_boston['BLK%'].values
y4 = df_boston['STL%'].values

fig = go.Figure()
fig.add_trace(go.Scatter(x=x, y=y1, marker = dict(color = '#00b159'), stackgroup='one', groupnorm='percent'))
fig.add_trace(go.Scatter(x=x, y=y2, marker = dict(color = '#ffc425'), stackgroup='one', groupnorm='percent'))
fig.add_trace(go.Scatter(x=x, y=y3, marker = dict(color = '#3498DB'), stackgroup='one', groupnorm='percent'))
fig.add_trace(go.Scatter(x=x, y=y4, marker = dict(color = '#ff3b30'), stackgroup='one', groupnorm='percent'))

fig.update_layout(width = 990, height = 650, xaxis_type='category',
                  yaxis=dict(range=[1, 100],nticksuffix='%'))

fig.show()

In [96]:
# Area Chart con relleno interior
x = df_boston.Player.values
y1 = df_boston['TRB%'].values
y2 = df_boston['AST%'].values

fig = go.Figure()

fig.add_trace(go.Scatter(x=x, y=y1, marker = dict(color = '#00b159'), fill = None ) )
fig.add_trace(go.Scatter(x=x, y=y2, fill='tonexty', marker = dict(color = '#ffc425')))

fig.update_layout(width = 980 , height = 600)
fig.show()

In [97]:
# Sunburst Chart con los minutos por posición y jugador
fig = px.sunburst(df_boston, path=['Pos' , 'Player'], values='MP')
fig.update_layout (height = 800 , width = 800)
fig.show()

In [98]:
# similar al anterior pero añadiendo el color por una variable
fig = px.sunburst(df_boston, path=['Pos' , 'Player'], values='MP', color= 'MP')
fig.update_layout (height = 800 , width = 800)
fig.show()

In [99]:
# similar al anterior 
fig = px.sunburst(df_boston, path=['Pos' , 'Player'], values='MP', color= 'MP',
                   color_continuous_scale=px.colors.sequential.Aggrnyl)

fig.update_layout(height = 800, width = 800)
fig.show()

In [100]:
# nos quedamos con menos atributos para simplificar visualizaciones
df_boston_2 = df_boston.iloc[:,0:10]
df_boston_2.head(5)

Player Pos  Age   Tm   G    MP   PER    TS%   3PAr    FTr
67     Jaylen Brown  SF   23  BOS  35  1169  18.0  0.598  0.359  0.302
139  Carsen Edwards  SG   21  BOS  27   267   6.5  0.432  0.621  0.137
149      Tacko Fall   C   24  BOS   4    21  24.1  0.690  0.000  0.273
185   Javonte Green  SF   26  BOS  32   248  14.0  0.623  0.235  0.221
205  Gordon Hayward  SF   29  BOS  24   759  18.4  0.596  0.319  0.141

In [101]:
# tabla básica
fig = go.Figure(data=[go.Table(header=dict(values=list(df_boston_2.columns)),
                            cells=dict(values=[
                                df_boston_2['Player'] , df_boston_2['Pos'] , df_boston_2['Age'] ,
                                df_boston_2['Tm'] , df_boston_2['G'], df_boston_2['MP'],
                                df_boston_2['PER'], df_boston_2['TS%'], df_boston_2['3PAr'],
                                df_boston_2['FTr']]))])

fig.show()

In [102]:
# tabla con otro estilo
fig = go.Figure(data=[go.Table(header=dict(values=list(df_boston_2.columns),
                                           fill_color='paleturquoise', align='left'),
                                 cells=dict(values=[df_boston_2['Player'] , df_boston_2['Pos'] , 
                                            df_boston_2['Age'] , df_boston_2['Tm'] , df_boston_2['G'],
                                            df_boston_2['MP'], df_boston_2['PER'], df_boston_2['TS%'],
                                            df_boston_2['3PAr'], df_boston_2['FTr']],
                                            fill_color='lavender', align='center'))])

fig.update_layout(width=990, height=350)
fig.show()

In [103]:
# diferentes estilos
fig = go.Figure(data=[go.Table(header=dict(values=list(df_boston_2.columns), line_color='black',  
                                              fill_color='orange', align='center', 
                                              height=40, font=dict(color='white', size=18)),
                                 cells=dict(values=[df_boston_2['Player'] , df_boston_2['Pos'] , 
                                                     df_boston_2['Age'] , df_boston_2['Tm'] , 
                                                     df_boston_2['G'], df_boston_2['MP'],
                                                     df_boston_2['PER'], df_boston_2['TS%'],
                                                     df_boston_2['3PAr'], df_boston_2['FTr']],
                                            line_color='darkgrey', fill_color='lightcyan', align='left'))])

fig.show()

In [104]:
# diferentes estilos
rowEvenColor = 'lightgrey'
rowOddColor = 'white'

fig = go.Figure(data=[go.Table(columnwidth = [80,80,80,80,80,80,120,80,80,80,120],
                                header=dict(values=['<b>Player</b>','<b>Pos</b>','<b>Age</b>',
                                                     '<b>Tm</b>','<b>G</b>','<b>MP</b>','<b>PER</b>',
                                                     '<b>TS%</b>','<b>3PAr</b>','<b>FTr</b>'],
                                              fill_color='#8BC34A', line = dict(color = '#689F38' , width = 4),
                                              align='center', font_size=12, font_color = 'white'),
                                 cells=dict(values=[df_boston_2['Player'] , df_boston_2['Pos'] , 
                                                     df_boston_2['Age'] , df_boston_2['Tm'] , 
                                                     df_boston_2['G'], df_boston_2['MP'],
                                                     df_boston_2['PER'], df_boston_2['TS%'],
                                                     df_boston_2['3PAr'], df_boston_2['FTr']],
                                             fill_color = [[rowOddColor,rowEvenColor]*5],
                                             line = dict(color = 'lightgreen' , width = 4),
                                             align ='center', font_size=12,
                                             font = dict(color = 'darkslategray', size = 11), height=40))])

fig.update_layout(width=990, height=500)
fig.show()

In [106]:
# tabla simple usando función create_table
fig = ff.create_table(df_boston_2.tail(5),height_constant=50)
fig.show()

In [107]:
# tabla simple con colores
colorscale = [[0, 'red'],[.5, '#DCE775'],[1, '#C0CA33']]

fig = ff.create_table(df_boston_2.tail(5), height_constant=50, colorscale=colorscale)

fig.show()

In [108]:
# Cambiando color de la fuente
colorscale = [[0, 'red'],[.5, '#DCE775'],[1, '#C0CA33']]
font=['white', '#212121' , 'red']

fig = ff.create_table(df_boston_2.tail(5), height_constant=50, colorscale=colorscale, font_colors=font)

fig.show()

In [110]:
# cambiando tamaño fuente
colorscale = [[0, 'red'],[.5, '#DCE775'],[1, '#C0CA33']]
font=['white', '#212121' , 'red']

fig = ff.create_table(df_boston_2.tail(5), height_constant=50, colorscale=colorscale, font_colors=font)

for i in range(len(fig.layout.annotations)):
    fig.layout.annotations[i].font.size = 10   

fig.show()

In [112]:
# tablas y gráficos
table_data = df_boston.loc[:, ['PER','WS','OBPM']].head(6)
fig = ff.create_table(table_data, height_constant=60)

fig.add_trace(go.Scatter(x = df_boston.Player.values, y = df_boston['PER'],
                    mode = 'lines', name = 'PER', xaxis='x2', yaxis='y2'))

fig.add_trace(go.Scatter(x = df_boston.Player.values, y = df_boston['WS'],
                    mode = 'lines', name = 'WS', xaxis='x2', yaxis='y2'))

fig.add_trace(go.Scatter(x = df_boston.Player.values, y = df_boston['OBPM'],
                    mode = 'lines', name = 'OBPM', xaxis='x2', yaxis='y2'))

fig.update_layout(title=dict(text = "PER-WS-OBPM",x=0.5,y=0.98), 
                    paper_bgcolor= '#dbdbdb', margin = {'t':50, 'b':100},
                    xaxis = {'domain': [0, .5] , 'title' : 'Stats'},
                    xaxis2 = {'domain': [0.6, 1.] , 'title' : 'Players'},
                    yaxis2 = {'anchor': 'x2', 'title': 'Count'},
                    width = 950, height = 600)

fig.show()

In [114]:
# similar a la anterior pero poniendo debajo la tabla y arriba el gráfico
table_data = df_boston.loc[:, ['PER', 'WS', 'OBPM', 'DBPM', 'OWS', 'AST%']].head(6)
fig = ff.create_table(table_data, height_constant=60)

fig.add_trace(go.Scatter(x = df_boston.Player.values, y = df_boston['PER'],
                    mode = 'lines', name = 'PER', xaxis='x2', yaxis='y2'))

fig.add_trace(go.Scatter(x = df_boston.Player.values, y = df_boston['WS'],
                    mode = 'lines', name = 'WS', xaxis='x2', yaxis='y2'))

fig.add_trace(go.Scatter(x = df_boston.Player.values, y = df_boston['OBPM'],
                    mode = 'lines', name = 'OBPM', xaxis='x2', yaxis='y2'))

fig.add_trace(go.Scatter(x = df_boston.Player.values, y = df_boston['DBPM'],
                    mode = 'lines', name = 'DBPM', xaxis='x2', yaxis='y2'))

fig.add_trace(go.Scatter(x = df_boston.Player.values, y = df_boston['OWS'],
                    mode = 'lines', name = 'OWS', xaxis='x2', yaxis='y2'))

fig.add_trace(go.Scatter(x = df_boston.Player.values, y = df_boston['AST%'],
                    mode = 'lines', name = 'AST%', xaxis='x2', yaxis='y2'))

fig.update_layout(title_text = 'Boston Celtics Stats', height = 800, margin = {'t':50, 'l':20},
    yaxis = {'domain': [0, .3]}, yaxis2 = {'domain': [.4, 1], 'anchor': 'x2', 'title': 'Values'},
    xaxis2 = {'anchor': 'y2'})

fig.show()

In [115]:
# ejemplo gráfica de burbujas
x = df_boston.Player
y = df_boston.PER

op = np.random.uniform(0.2, 1 ,9)

data = go.Scatter(x = x, y = y, mode = 'markers', marker = dict(size = y,opacity = op))

fig = go.Figure(data=data)
fig.show()

In [116]:
# gráfico de burbujas usando plotly express
fig = px.scatter(df_boston, x = "Player",  y = "PER", size = "PER", 
                 color = "Pos", hover_name = "Age", size_max = 60)

fig.show()